File: 01-backend-shootout/servers/go/main.go
package main

import (
	"bytes"
	"compress/zlib"
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"math/rand"
	"net/http"
	"strconv"
	"strings"
)

type User struct {
	ID    int    `json:"id"`
	Name  string `json:"name"`
	Email string `json:"email"`
}

var users = make(map[int]User)
var payload = bytes.Repeat([]byte("x"), 10000)

func init() {
	for i := 1; i <= 1000; i++ {
		users[i] = User{ID: i, Name: fmt.Sprintf("User %d", i), Email: fmt.Sprintf("user%d@example.com", i)}
	}
}

func writeJSON(w http.ResponseWriter, v any) {
	w.Header().Set("Content-Type", "application/json")
	json.NewEncoder(w).Encode(v)
}

func main() {
	http.HandleFunc("/hello", func(w http.ResponseWriter, r *http.Request) {
		writeJSON(w, map[string]string{"message": "Hello, World!"})
	})

	http.HandleFunc("/user/", func(w http.ResponseWriter, r *http.Request) {
		parts := strings.Split(r.URL.Path, "/")
		if len(parts) < 3 {
			http.NotFound(w, r)
			return
		}
		id, err := strconv.Atoi(parts[2])
		if err != nil {
			http.NotFound(w, r)
			return
		}
		u, ok := users[id]
		if !ok {
			writeJSON(w, map[string]string{"error": "not found"})
			return
		}
		writeJSON(w, u)
	})

	http.HandleFunc("/cpu", func(w http.ResponseWriter, r *http.Request) {
		// Hashing
		h := sha256.Sum256(payload)
		// JSON processing
		type Item struct {
			ID  int     `json:"id"`
			Val float64 `json:"val"`
		}
		items := make([]Item, 100)
		for i := range items {
			items[i] = Item{ID: i, Val: rand.Float64()}
		}
		data := map[string]any{"items": items}
		b, _ := json.Marshal(data)
		var parsed map[string]any
		json.Unmarshal(b, &parsed)
		// Compression
		var buf bytes.Buffer
		zw := zlib.NewWriter(&buf)
		zw.Write(payload)
		zw.Close()

		writeJSON(w, map[string]any{
			"hash":            fmt.Sprintf("%x", h),
			"items":           100,
			"compressed_size": buf.Len(),
		})
	})

	fmt.Println("Go server running on port 3000")
	http.ListenAndServe(":3000", nil)
}

File: 01-backend-shootout/servers/rust/src/main.rs
use axum::{extract::Path, response::Json, routing::get, Router};
use flate2::{write::ZlibEncoder, Compression};
use rand::Rng;
use serde_json::{json, Value};
use sha2::{Digest, Sha256};
use std::collections::HashMap;
use std::io::Write;
use std::sync::Arc;

type Users = Arc<HashMap<u32, Value>>;

fn make_users() -> Users {
    let mut m = HashMap::new();
    for i in 1u32..=1000 {
        m.insert(
            i,
            json!({"id": i, "name": format!("User {}", i), "email": format!("user{}@example.com", i)}),
        );
    }
    Arc::new(m)
}

#[tokio::main]
async fn main() {
    let users = make_users();

    let app = Router::new()
        .route("/hello", get(hello))
        .route(
            "/user/:id",
            get({
                let users = users.clone();
                move |Path(id): Path<u32>| {
                    let users = users.clone();
                    async move {
                        let val = users
                            .get(&id)
                            .cloned()
                            .unwrap_or_else(|| json!({"error": "not found"}));
                        Json(val)
                    }
                }
            }),
        )
        .route("/cpu", get(cpu));

    println!("Rust (Axum) server running on port 3000");
    let listener = tokio::net::TcpListener::bind("0.0.0.0:3000").await.unwrap();
    axum::serve(listener, app).await.unwrap();
}

async fn hello() -> Json<Value> {
    Json(json!({"message": "Hello, World!"}))
}

async fn cpu() -> Json<Value> {
    let payload = vec![b'x'; 10000];
    // Hash
    let mut hasher = Sha256::new();
    hasher.update(&payload);
    let hash = format!("{:x}", hasher.finalize());
    // JSON
    let mut rng = rand::thread_rng();
    let items: Vec<Value> = (0..100)
        .map(|i| json!({"id": i, "val": rng.gen::<f64>()}))
        .collect();
    let data = json!({"items": items});
    let serialized = serde_json::to_string(&data).unwrap();
    let parsed: Value = serde_json::from_str(&serialized).unwrap();
    let item_count = parsed["items"].as_array().unwrap().len();
    // Compress
    let mut encoder = ZlibEncoder::new(Vec::new(), Compression::default());
    encoder.write_all(&payload).unwrap();
    let compressed = encoder.finish().unwrap();

    Json(json!({
        "hash": hash,
        "items": item_count,
        "compressed_size": compressed.len()
    }))
}
